File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        seen = set()
        curr = head
        while curr:
          if curr in seen:
            return curr
          
          seen.add(curr)
          curr = curr.next
        return None

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var detectCycle = function(head) {
  let slow = head, fast = head;
  
  while (fast && fast.next && fast.next.next) {
    // find intersection point of cycle
    slow = slow.next;
    fast = fast.next.next;
    
    if (slow === fast) {
      // find starting point of cycle
      slow = head;
      while (slow !== fast) {
        slow = slow.next;
        fast = fast.next;
      }
      return fast;
    }
  }
  
  return null;
};